artiq/device_db_testing/eems.py

```diff
@@ -0,0 +1,146 @@
+from migen import *
+from migen.build.generic_platform import *
+from migen.genlib.io import DifferentialOutput
+
+from artiq.gateware import rtio
+from artiq.gateware.rtio.phy import ttl_simple, ttl_serdes_7series, spi2
+from artiq.build_soc import build_artiq_soc
+from artiq import __version__ as artiq_version
+
+from artiq.gateware.targets.kasli import (_dio, _urukul)
+
+import json
+
+#Create PHYs and rtio channels for a dio device and return dictionary with information
+#about those channels
+def eem_dio(cls, eem, name):
+    info = {
+        name: {
+            "type": "dio",
+            "location": (eem,),
+            "rtio_channels": {}
+        }
+    }
+
+    cls.platform.add_extension(_dio(eem))
+    for port in range(8):
+        pads = cls.platform.request(eem, port)
+        phy = ttl_serdes_7series.InOut_8X(pads.p, pads.n)
+        cls.submodules += phy
+        info[name]["rtio_channels"].update(
+            {"ttl"+str(port): {
+                            "channel_object": rtio.Channel.from_phy(phy),
+                            "class": check_phy_class(phy),
+                            "channel_num": None
+                            }
+            }
+        )
+
+    return info
+
+
+#Create PHYs and rtio channels  for an urukul device and return dictionary with information
+#about those channels
+def eem_urukul(cls, eem, eem_aux, name):
+    info = {

```

Comment:
It might be better to push all this info down into the PHYs themselves and introduce other classes that maintain the rest of the metadata that spans multiple PHYs.
If you really need thi dict-of-dict-and-lists, then at least document it.